### PR TITLE
Fix found bugs in js-codegen 

### PR DESCRIPTION
**Pull request by jarivm**
Some bugs were found/fixed:

- [x] Check if a context node is truthy, becuase `domFacade` can return null.
- [x] Attribute axis without attributes can crash. This could happen because `domFacade` can return `null` when using `getAllAttributes`.
- [x] `rootExpr` is not handled according to the spec. This should be fixed by finding the root for the given context (if expression is absolute).

### CODE REVIEW

**Comment by DrRataplan** [comment]
## BundleMon

  
  
  
  
  
<details open>
<summary>Files updated (2)</summary>

Status | Path | Size | Limits
:------------: | ------------ | :------------: | :------------:
:white_check_mark: | fontoxpath.esm.js<br/> | 90.97KB (+50B +0.05%) | -
:white_check_mark: | fontoxpath.js<br/> | 90.85KB (+50B +0.05%) | -
</details>

  
  


Total files change +100B +0.05%

Final result: :white_check_mark:

---
<p align="center"><a href="https://bundlemon.now.sh/projects/609024effb87a20009a6b096/reports?branch=fix-codegen&resolution=all" target="_blank" rel="noreferrer noopener">Current branch size history</a> | <a href="https://bundlemon.now.sh/projects/609024effb87a20009a6b096/reports?branch=master&resolution=days" target="_blank" rel="noreferrer noopener">Target branch size history</a></p>

File: src/jsCodegen/emitStep.ts [hunk 1]
```diff
@@ -90,8 +97,18 @@ function emitParentAxis(
 	return emitSingleNodeAxis(test, predicates, nestLevel, nestedCode, contextNodeCode);
 }
 
-function formatConditionCode(condition: string) {
-	return condition.length !== 0 ? `&& ${condition}` : '';
+function formatConditions(...conditions: string[]) {
+	return conditions.reduce((formatted, current, index) => {
```
Comment (DrRataplan) [inline src/jsCodegen/emitStep.ts:101]: aka `conditions.join(' && ')` ?

File: src/jsCodegen/emitStep.ts [hunk 2]
```diff
@@ -130,13 +154,16 @@ function emitSingleNodeAxis(
 	nestedCode: string,
 	contextNodeCode: string
 ): PartialCompilationResult {
-	const testEvaluatationCode = formatConditionCode(test);
-	const predicateEvaluationCode = formatConditionCode(predicates);
+	const isFirstPassCode = `i${nestLevel} == 0`;
```
Comment (DrRataplan) [inline src/jsCodegen/emitStep.ts:157]: why eqeq? why not eqeqeq?

File: test/specs/jsCodegen/axes.tests.ts [hunk 3]
```diff
@@ -25,21 +30,35 @@ describe('axes (js-codegen)', () => {
...
-			evaluateXPathWithCodegen('/xml/@id', documentNode, null, ReturnType.BOOLEAN)
+			evaluateXPathWithJsCodegen('/xml/@id', documentNode, null, ReturnType.BOOLEAN)
+		);
+	});
+
+	it('evaluates with the attribute axis when there are no attributes', () => {
+		documentNode = new slimdom.Document();
+		jsonMlMapper.parse(['xml', 'Some text'], documentNode);
+
+		// Type is "any" to be able to assign null to the attributes property.
```
Comment (DrRataplan) [inline test/specs/jsCodegen/axes.tests.ts:51]: I don't get this comment. Care to elaborate?

File: test/specs/jsCodegen/nodeTests.tests.ts [hunk 4]
```diff
@@ -91,4 +91,15 @@ describe('kind tests', () => {
 			documentNode.firstChild
 		);
 	});
+
+	it('handles detached nodes', () => {
```
Comment (DrRataplan) [inline test/specs/jsCodegen/nodeTests.tests.ts:95]: Do we need to go over the other paths-that-can-return-null as well? so child:: w/o children, attribute w/o matching attributes, etcetera?

**Review by DrRataplan — CHANGES_REQUESTED** [review]

File: src/jsCodegen/emitStep.ts (hunk 1 above)
Comment (jarivm) [reply src/jsCodegen/emitStep.ts:101]: That + `filter`. Thanks!

File: test/specs/jsCodegen/nodeTests.tests.ts (hunk 4 above)
Comment (jarivm) [reply test/specs/jsCodegen/nodeTests.tests.ts:95]: Yes. This is the cause for other errors.

Comment (DrRataplan) [reply test/specs/jsCodegen/nodeTests.tests.ts]: As long as you're not using firstchild/lastchild/sibling axes. :+1:

**Review by DrRataplan — APPROVED** [review]

**Pull request merged**